### PR TITLE
Introduce CodeSectionINTEL storage class

### PR DESCRIPTION
**Pull request by MrSidims**
This storage class is used for function pointers. It's added as based on cl_intel_function_pointers specification, it is not guaranteed that sizeof(void(*)(void) == sizeof(void *) - to allow consumers use this fact, we cannot say that function pointer belongs to the same storage class as data pointers.

It wasn't added during initial implementation, now it's time to fill this gap. As it would be a breaking change its generation is added only under -spirv-emit-function-ptr-addr-space option. Also SPIR-V consumer may pass this option during reverse translation to get new address space even in a case, when OpConstantFunctionPointerINTEL doesn't reside in CodeSectionINTEL storage class.

Expected behavior:
- No option is passed to the forward translation stage and function pointers are in addrspace(9):
   no CodeSectionINTEL storage class in SPIR-V
- The option is passed to the forward translation stage and function pointers are in addrepace(9):
   CodeSectionINTEL storage class is generated
- No option is passed to the reverse translation stage: function pointers are in private address space
- The option is passed to the reverse translation stage: function pointers are in addrspace(9)

Spec:
https://github.com/intel/llvm/blob/sycl/sycl/doc/design/spirv-extensions/SPV_INTEL_function_pointers.asciidoc

The previous approach: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1392

### CODE REVIEW

File: test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/alias.ll [hunk 1]
```diff
@@ -0,0 +1,49 @@
+; RUN: llvm-as %s -o %t.bc
```
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/alias.ll:1]: Note: most of these tests are copies from SPV_INTEL_function_pointers/* tests with -spirv-emit-function-ptr-addr-space option added in reverse translation to check if addrspace(9) is generated. The only exception to this is function-pointer-dedicated-as.ll test

File: test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll [hunk 2]
```diff
@@ -0,0 +1,102 @@
+; This test checks how a function pointer in a dedicated addr space would be
+; translated with and without -spirv-emit-function-ptr-addr-space option.
+; Expected behaviour:
+; No option is passed to the forward translation stage - no CodeSectionINTEL storage class in SPIR-V
+; The option is passed to the forward translation stage - CodeSectionINTEL storage class is generated
+; No option is passed to the reverse translation stage - function pointers are in private address space
+; The option is passed to the reverse tranlsation stage - function pointers are in addrspace(9)
```
Comment (maarquitos14) [inline test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll:7]:
```suggestion
; The option is passed to the reverse translation stage - function pointers are in addrspace(9)
```

File: test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+; RUN: FileCheck < %t.spt %s --check-prefix=CHECK-SPIRV-NO-AS
+; RUN: llvm-spirv %t.bc --spirv-ext=+SPV_INTEL_function_pointers -o %t.spv
+; RUN: llvm-spirv -r -spirv-emit-function-ptr-addr-space %t.spv -o %t.r.bc
+; RUN: llvm-dis %t.r.bc -o %t.r.ll
+; RUN: FileCheck < %t.r.ll %s --check-prefix=CHECK-LLVM-AS
+
+; CHECK-SPIRV-AS: TypePointer [[#PtrTy:]] 5605 [[#]]
+; CHECK-SPIRV-AS: ConstantFunctionPointerINTEL [[#PtrTy]]
+
+; CHECK-SPIRV-NO-AS-NOT: TypePointer [[#]] 5605 [[#]]
```
Comment (maarquitos14) [inline test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll:46]: Is this enough or should we also check there's no `ConstantFunctionPointerINTEL`?

Comment (MrSidims) [reply test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll:46]: ConstantFunctionPointerINTEL is a must regardless of the option. But it's storage class is being changed from CodeSectionINTEL to Private and vice versa.

Comment (MrSidims) [reply test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll:46]: I should probably extend checks for CHECK-SPIRV-AS to show exact SPIR-V that is being generated when the option is passed. Yet IMHO a single check for CHECK-SPIRV-NO-AS is sufficient

Comment (maarquitos14) [reply test/extensions/INTEL/SPV_INTEL_function_pointers/CodeSectionINTEL/function-pointer-dedicated-as.ll:46]: Agreed.

**Review by maarquitos14 — APPROVED** [review]

**Comment by MrSidims** [comment]
@svenvh hi Sven, it's an Intel-related PR, but it adds a new option to the translator and allocates a new value for address spaces, so may I ask you to have a quick look at these changes?

**Comment by svenvh** [comment]
> @svenvh hi Sven, it's an Intel-related PR, but it adds a new option to the translator and allocates a new value for address spaces, so may I ask you to have a quick look at these changes?

Thanks for the ping; no concerns from my side for adding the new option.

**Review by VyacheslavLevytskyy — APPROVED** [review]

**Pull request merged**